Add instructions for existing AWS infrastructure (#1560)
kubernetes-sigs/cluster-api-provider-aws: docs/existing-aws-infrastructure.md

# Consuming Existing AWS Infrastructure

Normally, Cluster API will create infrastructure on AWS when standing up a new workload cluster. However, it is possible to have Cluster API re-use existing AWS infrastructure instead of creating its own infrastructure. Follow the instructions below to configure Cluster API to consume existing AWS infrastructure.

## Prerequisites

In order to have Cluster API consume existing AWS infrastructure, you will need to have already created the following resources:

* A VPC

> * One or more private subnets (subnets that do not automatically assign a public IP address to EC2 instances)

Reviewer: What CAPA considers public is mainly whether the subnet in question has a route to an internet gateway – if yes, it's public, otherwise it's private.

Author: Added some text to that effect via 38ce860

* A public subnet in the same Availability Zone (AZ) for each private subnet (this is required for NAT gateways to function properly)
* A NAT gateway for each private subnet, along with associated Elastic IP addresses
* An Internet gateway for all public subnets

> * Route table associations that provide connectivity to the Internet through a NAT gateway (for private subnets) or the Internet gateway (for public subnets)

Reviewer: It might be worth mentioning that that the public subnet / IGW bits are still necessary, even if you also specify an `internal` scheme for your [apiserver's load balancer](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/api/v1alpha3/types.go#L121). It looks like we're [unwilling to proceed without one](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/0b83f06314410a9a3d628f006a647086f51736a8/pkg/cloud/services/ec2/subnets.go#L67).

Reviewer: Actually, we do [defer to the public role tag](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/0b83f06314410a9a3d628f006a647086f51736a8/pkg/cloud/services/ec2/subnets.go#L210-L212), so if someone _really_ wants to they could tag an actually-private subnet as role "public" to get past that linked error.

Probably best to leave that out of the documentation, though, as I've no idea what might fail down the line if someone did such a thing.

Author: Updated via 38ce860. I did not add information specific to the requirement of a public subnet and IGW with regard to the API server load balancer, as those items are already needed anyway. If desired, I can add those details.

Author: I didn't specifically address this comment (the text already noted that the public subnet and IGW are required), but I can add some additional text if desired.

You will need the following information, which can be obtained either via the AWS Management Console or using the AWS CLI:

1. The ID of the VPC

> 2. The IDs of the subnets in the VPC that you want Cluster API to use

Reviewer: The subnet IDs are actually ignored; given a VPC we'll always look up all the subnets and populate them on the cluster object via [describe](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/0b83f06314410a9a3d628f006a647086f51736a8/pkg/cloud/services/ec2/subnets.go#L170).

Providing the subnet IDs wouldn't bother CAPA, but it might lead the humans to believe they can ask CAPA to use a subset of IDs by providing a reduced list. At least at present, CAPA would ignore that intent.

Author: Fixed via 38ce860.

Reviewer 2: TIL, and this seems...not ideal?

Author: I’ve tested it, and the subnet IDs are not required. I believe that text has been removed in the most recent commit against this PR.

Note that there is no need to create an Elastic Load Balancer (ELB), security groups, or EC2 instances; Cluster API will take care of these items.

> ## Tagging AWS Resources

Reviewer 3: I wouldn't expect the tags to be required if specifying vpc ids and subnet ids, but I'll try to verify.

Reviewer: Agreed – they're not required, and in fact not a good idea. The way CAPA knows whether it can modify the VPC/subnets/route tables/nat gateways/IGWs is by whether the VPC has the "owned" tag. If we mention the tags at all in this page, I would suggest saying something to the effect of "setting this tag will tell CAPA to take ownership over and modify the existing resources" and leaving it at that.

Author: References to CAPI-specific tags removed via 38ce860.

Author: Resolved via 38ce860, where I removed references to CAPI-specific tags being required.

* Route tables
* Internet gateway
* NAT gateways
* Elastic IPs assigned to NAT gateways

For the `sigs.k8s.io/cluster-api-provider-aws/cluster/<cluster-name>` tag, a value of "owned" is used by Cluster API itself to indicate resources are owned by a specific cluster, and that the lifecycle of the resource is tied to the lifecycle of the cluster. Cluster API recognizes a value of "shared" to indicate a resource may be shared between multiple clusters, and the lifecycle of said resource is not tied to any particular cluster. In this situation---where AWS infrastructure is _not_ being managed by Cluster API---the value should be something **other** than "owned". Using "shared" is acceptable.

For the `sigs.k8s.io/cluster-api-provider-aws/role` tag, Cluster API uses a variety of values to indicate resources are associated with a particular role. Here are the tag values that should be used:

* VPC, route tables, Internet gateway, and NAT gateways should use the value "common".
* Public subnets should use the value "public".
* Private subnets should use the value "private".
* Elastic IPs (for use by the NAT gateways) should use the value "apiserver".

> Additionally, for proper operation of the AWS cloud provider, all subnets where Kubernetes nodes reside should also have the `kubernetes.io/cluster/<cluster-name>` tag present.

Reviewer: Yes, and additionally private subnets should be tagged with `kubernetes.io/role/internal-elb=1` and public with `kubernetes.io/role/elb=1`, else the cloud provider can't find any subnets to create load balancers in.

Author: Added information about these tags via 38ce860.

## Configuring the AWSCluster Specification

Specifying existing infrastructure for Cluster API to use takes place in the specification for the AWSCluster object. Specifically, you will need to add an entry with the VPC ID and the IDs of all applicable subnets into the `networkSpec` field. Here is an example:

```yaml
spec:
  networkSpec:
    vpc:
      id: vpc-0425c335226437144
    subnets:
      - id: subnet-07758a9bc904d06af
      - id: subnet-0a3507a5ad2c5c8c3
      - id: subnet-02ad6429dd0532452
      - id: subnet-02b300779e9d895cf
      - id: subnet-03d8f353b289b025f

>       - id: subnet-0a2fe03b0d88fa078

Reviewer: As mentioned above, it's not necessary to include the subnets here. However, we do find it useful to include a default `sshKeyName` and `region` at this point, like so:

```
spec:
  networkSpec:
    vpc:
      id: vpc-0425c335226437144
  region: us-west-2
  sshKeyName: some-ssh-key
```

Though neither are required.

Author: Removed references to the list of subnets via 38ce860.

```

When you use `kubectl apply` to apply the Cluster and AWSCluster specifications to the management cluster, Cluster API will use the specified VPC ID and specified subnet IDs, and will not create a new VPC, new subnets, or other associated resources. It _will_, however, create a new ELB and new security groups.

## Configuring the AWSMachine Specification (Optional)

To distribute EC2 instances across different subnets or different AZs, you can add information to the AWSMachine specification. This is optional; without it, Cluster API will deploy to the first private subnet it discovers.

To specify that an EC2 instance should be placed in a specific subnet, add this to the AWSMachine specification:

```yaml
spec:
  subnet:
    id: subnet-0a3507a5ad2c5c8c3
```

To tell Cluster API that an EC2 instance should be placed in a particular AZ but allow Cluster API to select which subnet in that AZ can be used, add this to the AWSMachine specification:

```yaml
spec:

>   availabilityZone: "us-west-2a"

Reviewer 3: This should be failureDomain for v1alpha3
```suggestion
  failureDomain: "us-west-2a"
```

Author: Fixed via 38ce860.

In order to have Cluster API consume existing AWS infrastructure, you will need to have already created the following resources:

* A VPC
* One or more private subnets (subnets that do not have a route to an Internet gateway)
* A public subnet in the same Availability Zone (AZ) for each private subnet (this is required for NAT gateways to function properly)
* A NAT gateway for each private subnet, along with associated Elastic IP addresses
* An Internet gateway for all public subnets
* Route table associations that provide connectivity to the Internet through a NAT gateway (for private subnets) or the Internet gateway (for public subnets)

> Note that a public subnet (and associated Internet gateway) are required for the control plane of the workload cluster.

Reviewer: I think I've mislead you – we insist that there _be_ public subnets, but it looks like we [know not to use them for the control plane load balancer ](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/9913ac8dbb96ed9bcffa6eba75b886cd4ef8c20b/pkg/cloud/services/elb/loadbalancer.go#L288) if the scheme is set to `internal`. Maybe:

```suggestion
Note that a public subnet (and associated Internet gateway) are required even if the control plane of the workload cluster is set to use an `internal` load balancer.
```

?

Author: I've adopted your change (included in 1667e3ab). Is it worthwhile including information on the mechanism by which an internal load balancer is configured, or is that information best suited to be elsewhere?

You will need the ID of the VPC that Cluster API should use. This information is available via the AWS Management Console or the AWS CLI. It is not necessary to have a list of the subnet IDs; Cluster API will discover the subnets if the VPC ID is provided.

Note that there is no need to create an Elastic Load Balancer (ELB), security groups, or EC2 instances; Cluster API will take care of these items.

## Tagging AWS Resources

> Cluster API itself does tag AWS resources it creates. The `sigs.k8s.io/cluster-api-provider-aws/cluster/<cluster-name>` (where `<cluster-name>` matches the `metadata.name` field of the Cluster object) tag, with a value of `owned`, tells Cluster API that it has ownership of the resource and is able to modify the resource and manage the lifecycle of the resource.

Reviewer: ```suggestion
Cluster API itself does tag AWS resources it creates. The tag `sigs.k8s.io/cluster-api-provider-aws/cluster/<cluster-name>` (where `<cluster-name>` matches the `metadata.name` field of the Cluster object), with a value of `owned`, tells Cluster API that it has ownership of the resource. In this case, Cluster API will modify and manage the lifecycle of the resource.
```

Author: 1667e3a includes this suggested wording change.

> When consuming existing AWS infrastructure, Cluster API does not require any tags to be present. The absence of the tags on an AWS resource indicates to Cluster API that it should not modify the resource or attempt to manage the lifecycle of the resource.

Reviewer: ```suggestion
When consuming existing AWS infrastructure, the Cluster API AWS provider does not require any tags to be present. The absence of the tags on an AWS resource indicates to Cluster API that it should not modify the resource or attempt to manage the lifecycle of the resource.
```

Author: I've adopted your change (included in 1667e3ab).

> However, the AWS cloud provider _does_ require certain tags in order to function properly. Specifically, all subnets where Kubernetes nodes reside should have the `kubernetes.io/cluster/<cluster-name>` tag present. Private subnets should also have the `kubernetes.io/role/internal-elb` tag with a value of 1, and public subnets should have the `kubernetes.io/role/elb` tag with a value of 1. These latter two tags help the cloud provider understand which subnets to use when creating load balancers.

Reviewer: ```suggestion
However, the built-in Kubernetes AWS cloud provider _does_ require certain tags in order to function properly. Specifically, all subnets where Kubernetes nodes reside should have the `kubernetes.io/cluster/<cluster-name>` tag present. Private subnets should also have the `kubernetes.io/role/internal-elb` tag with a value of 1, and public subnets should have the `kubernetes.io/role/elb` tag with a value of 1. These latter two tags help the cloud provider understand which subnets to use when creating load balancers.
```

Author: I've adopted your change (included in 1667e3ab).

    id: subnet-0a3507a5ad2c5c8c3
```

When using MachineDeployments, users can control subnet selection by adding information to the AWSMachineTemplate associated with that MachineDeployment, like this:

```yaml
spec:
  template:
    spec:
      subnet:
        id: subnet-0a3507a5ad2c5c8c3
```

> Users may either specify `failureDomain` on the Machine or MachineDeployment objects, _or_ users may explicitly specify subnet IDs on the AWSMachine or AWSMachineTemplate objects. The two settings are mutually exclusive.

Reviewer: ```suggestion
Users may either specify `failureDomain` on the Machine or MachineDeployment objects, _or_ users may explicitly specify subnet IDs on the AWSMachine or AWSMachineTemplate objects. If both are specified, the subnet ID is used and the `failureDomain` is ignored.
```

We aren't able to validate cross-object, so we can't prevent a determined user from setting up the scenario where both are specified. Better to be explicit about what happens in that case IMO.

Author: This suggested change is included in 1667e3a.
